code/hash/Hacl.Hash.SHA2.fst

```diff
@@ -1,45 +1,380 @@
 module Hacl.Hash.SHA2
 
+module ST = FStar.HyperStack.ST
+module HS = FStar.HyperStack
+
+module B = LowStar.Buffer
+
 open Hacl.Hash.Definitions
 open Spec.Hash.Definitions
 
-include Hacl.Hash.Core.SHA2
+open Lib.MultiBuffer
+open Lib.NTuple
+
+module Vec = Hacl.Spec.SHA2.Vec
+
+// To prove we are properly defining init
+friend Spec.Agile.Hash
+// To prove we are properly defining update_last
+friend Spec.Hash.Incremental
+// To know the definition of init
+friend Spec.SHA2
+
+#set-options "--fuel 0 --ifuel 0 --z3rlimit 100"
+
+inline_for_extraction noextract
+let mb_state_32 a = Hacl.Impl.SHA2.Core.state_t a Hacl.Spec.SHA2.Vec.M32
+
+#push-options "--ifuel 1"
+
+inline_for_extraction noextract
+let coerce_to_state (a:sha2_alg) (b:mb_state_32 a) : state (| a, () |)
+  = Lib.IntVector.reveal_vec_1 (word_t a);
+    b
+
+inline_for_extraction noextract
+let coerce_to_mb_state (a:sha2_alg) (b:state (| a, () |)) : mb_state_32 a
+  = Lib.IntVector.reveal_vec_1 (word_t a);
+    b
+
+val state_spec_v_lemma (a:sha2_alg) (st:Vec.state_spec a Vec.M32) : Lemma
+  (Lib.IntVector.reveal_vec_1 (word_t a);
+   st `Seq.equal` Lib.Sequence.index (Vec.state_spec_v st) 0)
+
+let state_spec_v_lemma a st =
+  let open Lib.Sequence in
+  let open Lib.IntVector in
+  reveal_vec_v_1 st.[0];
+  reveal_vec_v_1 st.[1];
+  reveal_vec_v_1 st.[2];
+  reveal_vec_v_1 st.[3];
+  reveal_vec_v_1 st.[4];
+  reveal_vec_v_1 st.[5];
+  reveal_vec_v_1 st.[6];
+  reveal_vec_v_1 st.[7];
+  reveal_vec_1 (word_t a);
+  eq_intro #(word a) #8 (Vec.state_spec_v st).[0] st
+
+let init_224 st =
```

Comment:
should we share the proofs between the four variants of `init`? same for `alloca`. 
should we move lemmas (including ones for `mk_update_multi`, `mk_update_last` and `mk_finish`) from this module to a standalone file?

Reply:
Good point thanks. I made generic `mk_init` and `mk_alloca`. Regarding the lemmas, it seems there only is `state_spec_v_lemma`. We could move it to another file, but since it's very specific to this correspondance, I'd be okay with keeping it here as well until we need it somewhere else. Wdyt?

Reply:
I meant moving the `calc` statements from code to lemmas. It is not a big deal but I prefer to separate proofs from stateful computations

Reply:
I'd usually be in favor, in this specific case though, I'm afraid about the amount of additional context that would need to be encoded in the Lemma (e.g., relations between `st`, `st'`, `blocks0` and `blocks0_multi`, ...). I'm happy to revisit later if it turns out to be unstable in the current state.